### ISSUE TITLE
[nixos] Error when pushing: ".git doesn't look like a git repository"

### ISSUE THREAD

**Issue by Derpford** [sublimehq/sublime_merge#742]
**Version info**

 - OS: NixOS 20.03
 - Build: 1116 Stable
 - home.nix: https://gist.github.com/Derpford/3f172e22f517aa7b061a24ca9494f23e (using [home-manager](https://github.com/rycee/home-manager))

**Description**

When pushing to remote, four error windows pop up stating that `.git doesn't look like a git repository`. I can still push to remote when using git from CLI.

**Steps to reproduce**

1. Open a repo that has a remote specified.
2. Commit something, making `local` ahead of `remote`.
3. Click on the "push" button (or hit Ctrl-Alt-Up)
4. Errors pop up.

**Expected behavior**

Sublime-merge prompts for a password, then pushes to remote.

**Comment by dpjohnst** [comment]
Hi @Derpford,

Sorry to hear you're having issues using Sublime Merge!

If possible, could you try updating to the latest development build of Sublime Merge and confirm if the problem persists?

Thanks,
\- Dylan

**Comment by BenjaminSchaaf** [comment]
This seems to be a nixos/home-manager issue. Sublime Merge provides an ssh-askpass executable in order to ask for credentials when pushing/pulling. This executable find the sublime_merge executable and runs it, passing along `argv[0]` as `ssh-askpass-sublime`. It seems like the links created by nixos/home-manager drop `argv[0]` on the floor, meaning that instead of spawning the credentials dialog box we try to launch another instance of Sublime Merge.
I'd suggest filing an issue with nixOS and/or contacting the maintainer of the Sublime Merge package. We don't provide an official distribution of Sublime Merge for nixOS.

**Comment by aiverson** [comment]
It looks like the problem is actually that the nix package for sublime-merge doesn't expose the ssh-askpass executable. The generated wrappers forward the argv[0] fine.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/sublime-merge/common.nix#L82

The nix build happens in two stages. First, the binaries are downloaded from the website, and patched to fix incorrectly hardcoded paths. In the second phase, a (possibly useless?) extra wrapper is created, some extra links are populated, a .desktop file is created, but the rest of the binaries aren't included for some reason.

So, in conclusion, the nix package retrieves, patches, and then discards the ssh-askpass executable, and sublime-merge fails to find the missing ssh-askpass executable and gives a misleading and incorrect error message because everything is broken.

@Derpford try adding lxqt.lxqt-openssh-askpass or x11_ssh_askpass to your config.nix and see if that makes it work. If it does, then the problem is that the nix package for sublime-merge doesn't include the ssh-askpass executable and providing one makes it work. If it doesn't, then sublime-merge doesn't launch ssh-askpass correctly.

**Comment by Derpford** [comment]
@aiverson Just added lxqt.lxqt-openssh-askpass to my home.nix, did a rebuild, and launched sublime-merge. Attempting a push creates the same set of error messages. Then I tried adding x11_ssh_askpass to my `with pkgs` block in my `/etc/nixos/configuration.nix` and did a rebuild, then tried again: same issue.

**Comment by BenjaminSchaaf** [comment]
Sublime Merge provides it's own ssh-askpass executable. It won't use the system ones. The correct executable is being run, ie. `ssh_askpass_sublime`, but it's either not begin forwarded the correct argv[0] or the `sublime_merge` it executes isn't being forwarded the correct argv[0].

**Comment by zookatron** [comment]
Hi all, I'm trying to determine what needs to be updated about the nix package in order to get this to work. I can say that the `ssh-askpass-sublime` binary IS being included by the nix package and that the `sublime_merge` executable is finding and executing it properly (I confirmed this with strace), however for whatever reason the `ssh-askpass-sublime` is just not creating the popup to ask for the passphrase and I don't understand why. When I try to run the `ssh-askpass-sublime` executable directly to test it out (passing the path to the git repo as argv[0]) I can never get it to actually show a popup, it just always spawns a new instance of `sublime_merge`.

Can you explain how the `ssh-askpass-sublime` executable has been modified from the standard X11 `ssh-askpass` executable (which I have no problems using)? Under what circumstances does it open the popup and under what circumstances does it try to spawn a new `sublime_merge`?

**Comment by BenjaminSchaaf** [comment]
The `ssh-askpass-sublime` executable only does two things: It looks for a `sublime_merge` executable in the same directory and runs it by passing through `argv[0]` such that `sublime_merge` is run with `argv[0] == "<somepath>/ssh-askpass-sublime"`. The `sublime_merge` executable then has a special-case where if the previous condition holds it will show a Sublime Merge styled ssh askpass popup.

**Comment by zookatron** [comment]
Thanks, that makes sense. However I can't quite figure out what is wrong about how `ssh-askpass-sublime` is passing the `<somepath>/ssh-askpass-sublime` path to `sublime_merge`. When I try running `sublime_merge` directly with that argument like `/install/path/sublime_merge /install/path/ssh-askpass-sublime` it says `Unable to find a git repository at /install/path/ssh-askpass-sublime`. Can you provide more details then about how exactly sublime_merge checks for the special `argv[0] == "<somepath>/ssh-askpass-sublime"` case? How exactly does it determine the `<somepath>` it checks for?

**Comment by aiverson** [comment]
How do you know the executable is being run? And more importantly, how is the executable being run when it isn't included in the nix package and so isn't on the path?

**Comment by BenjaminSchaaf** [comment]
`/install/path/sublime_merge /install/path/ssh-askpass-sublime` doesn't pass `ssh-askpass-sublime` as `argv[0]`. The equivalent would be: `bash -c "exec -a /install/path/ssh-askpass-sublime /install/path/sublime_merge"`. Sublime Merge ignores all but the leaf part for it's `argv[0]` check, as long as it ends with `/ssh-askpass-sublime` it will behave as askpass.

**Comment by zookatron** [comment]
Thanks for the clarification, for some reason I was confusing `argv[0]` with `argv[1]`. I'm now able to test out the popup by running that `exec -a` command. I was able to figure out why this is not working when running the `sublime_merge` in the package normally, it is because the `sublime_merge` that `ssh-askpass-sublime` is executing is actually a wrapper bash script that calls `exec` to run the actual binary after setting up the proper environment variables. There doesn't appear to be any way for the bash script to detect the `argv[0]` that was passed to bash as `$0` is overwritten with the path to the bash script. I'm not sure how we will resolve this issue with the package but we should be able to do something now that we understand what the problem is, thanks for your help @BenjaminSchaaf !

**Comment by BenjaminSchaaf** [comment]
Seeing as this is a nixOS/home-manager issue, I'll close this.
